pkg/rslog/buf_logger.go

```diff
@@ -0,0 +1,100 @@
+package rslog
+
+// Copyright (C) 2022 by RStudio, PBC.
+
+type BufLogEntry struct {
+	Level   LogLevel
+	Message string
+	Args    []interface{}
+	Logger  CoreLoggerImpl
+}
+
+type BufStorage struct {
+	Logs []BufLogEntry
+}
+
+type BufLogger struct {
+	CoreLogger CoreLoggerImpl
+	Storage    *BufStorage
+	flush      func()
+}
+
+var _ CoreLoggerImpl = new(BufLogger)
+
+func NewBufLogger(coreLogger CoreLoggerImpl, flush func()) *BufLogger {
+	return &BufLogger{
+		Storage: &BufStorage{
+			Logs: make([]BufLogEntry, 0),
+		},
+		CoreLogger: coreLogger,
+		flush:      flush,
+	}
+}
+
+// child creates and registers a wrapped logger implementation and returns a new buffered logger
+// that will pass down the same storage and fallback fields for its children, if any.
+func (buf *BufLogger) child(coreLogger CoreLoggerImpl) *BufLogger {
+	return &BufLogger{
+		CoreLogger: coreLogger,
+		Storage:    buf.Storage,
+	}
+}
+
+func (buf *BufLogger) Tracef(msg string, args ...interface{}) {
+	buf.Storage.Logs = append(buf.Storage.Logs, BufLogEntry{
+		Level:   TraceLevel,
+		Message: msg,
+		Args:    args,
+		Logger:  buf.CoreLogger,
+	})
```

Comment:
WDYT about having a `Storage.Add()` method to simplify the append in log methods?

Reply:
Also, there is risk of having a race condition with this multiple `append` calls. You could implement a single mutex lock in the proposed `Storage.Add()` method.